packages/core/src/middleware/koa-quota-guard.ts

```diff
@@ -0,0 +1,74 @@
+import type { MiddlewareType } from 'koa';
+
+import { EnvSet } from '#src/env-set/index.js';
+import RequestError from '#src/errors/RequestError/index.js';
+import { type CloudConnectionLibrary } from '#src/libraries/cloud-connection.js';
+import type Queries from '#src/tenants/Queries.js';
+import assertThat from '#src/utils/assert-that.js';
+import { getTenantSubscriptionPlan } from '#src/utils/subscription/index.js';
+import { type SubscriptionQuota } from '#src/utils/subscription/types.js';
+
+type UsageGuardConfig = {
+  key: keyof SubscriptionQuota;
+  cloudConnection: CloudConnectionLibrary;
+  queries: Queries;
+};
+
+const getTenantUsage = async (key: keyof SubscriptionQuota, queries: Queries): Promise<number> => {
+  if (key === 'applicationsLimit') {
+    return queries.applications.countNonM2MApplications();
+  }
+
+  // TODO: add other keys
+
+  throw new Error('Unsupported subscription quota key');
+};
+
+export default function koaQuotaGuard<StateT, ContextT, ResponseBodyT>({
+  key,
+  queries,
+  cloudConnection,
+}: UsageGuardConfig): MiddlewareType<StateT, ContextT, ResponseBodyT> {
+  return async (ctx, next) => {
+    const { isCloud, isIntegrationTest, isProduction } = EnvSet.values;
+
+    // Disable in production until price feature is ready
+    if (!isCloud || isIntegrationTest || isProduction) {
+      return next();
+    }
+
+    const plan = await getTenantSubscriptionPlan(cloudConnection);
+    const limit = plan.quota[key];
+
+    if (typeof limit === 'boolean') {
+      assertThat(
+        limit,
+        new RequestError({
+          code: 'subscription.limit_exceeded',
+          status: 403,
+          data: {
+            key,
+          },
+        })
+      );
+    } else if (typeof limit === 'number') {
+      const tenantUsage = await getTenantUsage(key, queries);
+
+      assertThat(
+        tenantUsage < limit,
```

Comment:
just to double check we do not allow the quota equals the limit?

Reply:
in the mock data, i see limits of 0, so the quota guard will always throw an error?

Reply:
Yes, because this guard is in the place where it is going to "add" one.

Reply:
maybe can add a comment to this middleware so that reviewers can clearly know how it works